File: algorithms/[695]岛屿的最大面积/solution.py
# topics = ["广度优先搜索"]

from collections import deque
from typing import List, Deque, Tuple


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0
        m = len(grid)
        n = len(grid[0])
        visited = [[False] * n for _ in range(m)]
        q: Deque[Tuple[int, int]] = deque()
        res = 0

        def bfs(i: int, j: int) -> int:
            q.append((i, j))
            visited[i][j] = True
            area = 0

            while q:
                x, y = q.popleft()
                area += 1
                for dx, dy in [[0, -1], [1, 0], [0, 1], [-1, 0]]:
                    nx = x + dx
                    ny = y + dy
                    if (
                        0 <= nx < m
                        and 0 <= ny < n
                        and grid[nx][ny]
                        and not visited[nx][ny]
                    ):
                        q.append((nx, ny))
                        visited[nx][ny] = True

            return area

        for i in range(m):
            for j in range(n):
                if grid[i][j] and not visited[i][j]:
                    res = max(res, bfs(i, j))
                    print(i, j, res)

        return res

File: algorithms/[695]岛屿的最大面积/solution.ts
// topics = ["深度优先搜索"]

// 递归实现
function maxAreaOfIsland(grid: number[][]): number {
  if (!grid.length) return 0;
  const m = grid.length;
  const n = grid[0].length;
  const visited = Array.from({ length: m }, () => new Array<boolean>(n).fill(false));
  let res = 0;

  const dfs = (x: number, y: number): number => {
    visited[x][y] = true;
    let area = 1;

    for (const [dx, dy] of [[0, -1], [1, 0], [0, 1], [-1, 0]]) {
      const nx = x + dx;
      const ny = y + dy;
      if (grid[nx]?.[ny] && !visited[nx][ny]) {
        area += dfs(nx, ny);
      }
    }

    return area;
  };

  for (let i = 0; i < m; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (grid[i][j] && !visited[i][j]) {
        res = Math.max(res, dfs(i, j));
      }
    }
  }

  return res;
}

// 栈实现
function maxAreaOfIsland2(grid: number[][]): number {
  if (!grid.length) return 0;
  const m = grid.length;
  const n = grid[0].length;
  const visited = Array.from({ length: m }, () => new Array<boolean>(n).fill(false));
  const stack: [number, number][] = [];
  let res = 0;

  const dfs = (i: number, j: number): number => {
    stack.push([i, j]);
    visited[i][j] = true;
    let area = 0;

    while (stack.length) {
      const [x, y] = stack.pop()!;
      area += 1;
      for (const [dx, dy] of [[0, -1], [1, 0], [0, 1], [-1, 0]]) {
        const nx = x + dx;
        const ny = y + dy;
        if (grid[nx]?.[ny] && !visited[nx][ny]) {
          stack.push([nx, ny]);
          visited[nx][ny] = true;
        }
      }
    }

    return area;
  };

  for (let i = 0; i < m; i += 1) {
    for (let j = 0; j < n; j += 1) {
      if (grid[i][j] && !visited[i][j]) {
        res = Math.max(res, dfs(i, j));
      }
    }
  }

  return res;
}
